lib/spack/llnl/util/tty/pty.py

```diff
@@ -0,0 +1,287 @@
+        return self.get_canon_echo_attrs() == (False, False)
+
+    def background(self):
+        """True if pgid is in a background pgroup of master_fd's terminal."""
+        return self.pgid != os.tcgetpgrp(self.master_fd)
+
+    def tstp(self):
+        """Send SIGTSTP to the controlled process."""
+        self.hline("tstp")
+        os.killpg(self.pgid, signal.SIGTSTP)
+
+    def cont(self):
+        self.hline("cont")
+        os.killpg(self.pgid, signal.SIGCONT)
+
+    def fg(self):
+        self.hline("fg")
+        os.tcsetpgrp(self.master_fd, os.getpgid(self.pid))
+
+    def bg(self):
+        self.hline("bg")
+        with log.ignore_signal(signal.SIGTTOU):
+            os.tcsetpgrp(self.master_fd, os.getpgrp())
+
+    def write(self, byte_string):
+        self.hline("write '%s'" % byte_string.decode("utf-8"))
+        os.write(self.master_fd, byte_string)
+
+    def wait(self, condition):
+        start = time.time()
+        while (((time.time() - start) < self.timeout) and not condition()):
+            time.sleep(1e-2)
+        assert condition()
+
+    def wait_enabled(self):
+        self.wait(lambda: self.input_on() and not self.background())
+
+    def wait_disabled(self):
+        self.wait(lambda: not self.input_on() and self.background())
+
+    def wait_disabled_fg(self):
+        self.wait(lambda: not self.input_on() and not self.background())
+
+    def proc_status(self):
+        status = self.ps("-p", str(self.pid), "-o", "stat", output=str)
+        status = re.split(r"\s+", status.strip(), re.M)
+        return status[1]
+
+    def wait_stopped(self):
+        self.wait(lambda: "T" in self.proc_status())
+
+    def wait_running(self):
+        status = self.proc_status()
+        self.wait(lambda: "T" not in status)
+
+
+def _child_process(
+        tty_name, stdout_fd, stderr_fd, ready, child_function, attrs):
+    """Child process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``child_function``.
+
+    """
+    # new process group, like a command or pipeline launched by a shell
+    os.setpgrp()
+
+    # take controlling terminal and set up pty IO
+    stdin_fd = os.open(tty_name, os.O_RDWR)
+    os.dup2(stdin_fd, sys.stdin.fileno())
+    os.dup2(stdout_fd, sys.stdout.fileno())
+    os.dup2(stderr_fd, sys.stderr.fileno())
+    os.close(stdin_fd)
+
+    if attrs.get("debug"):
+        sys.stderr.write("child: stdin.isatty(): %s\n" % sys.stdin.isatty())
+
+    # tell the parent that we're really running
+    if attrs.get("debug"):
+        sys.stderr.write("child: ready!\n")
+    ready.value = True
+
+    child_function(attrs)
+
+
+def _master_process(master_function, child_function, attrs):
+    """Master process wrapper for PseudoShell.
+
+    Handles the mechanics of setting up a PTY, then calls
+    ``master_function``.
+
+    """
+    os.setsid()   # new session; this process is the controller
+
+    master_fd, child_fd = os.openpty()
```

Comment:
Note that https://docs.python.org/3/library/os.html (specifically the section on `openpty`) mentions that these are non-inheritable, which is a problem on Windows (and probably also Python >= 3.8 on Mac OS). The doc suggests that the `pty `module may be more portable, but I don't know if that is useful here.

Reply:
They're not actually passed to the child process, so I think this is safe.